src/sentry/api/endpoints/organization_stats_v2.py

```diff
@@ -0,0 +1,239 @@
+
+    def zerofill(self, start, end, rollup):
+        self.errors = zerofill(self.errors, start, end, rollup, "time")
+        self.transactions = zerofill(self.transactions, start, end, rollup, "time")
+        self.attachments = zerofill(self.attachments, start, end, rollup, "time")
+
+    def build_fields(self):
+        return {CATEGORY_NAME_MAP[category]: values.serialize() for category, values in self}
+
+
+# TODO: perhaps remove this class
+class TimeSeriesValues:
+    def __init__(self, start, end, rollup):
+        self.values = zerofill({}, start, end, rollup, "time")
+
+    def update(self, row):
+        self.values[row["time"]].update(row)
+
+    def serialize(self):
+        return [value.serialize() for value in self.values.values()]
+
+
+class StatMeasure:
+    def __init__(self, quantity=0, times_seen=0):
+        self.quantity = quantity
+        self.times_seen = times_seen
+
+
+class MeasureValue:
+    def __init__(self, time):
+        self.accepted = StatMeasure(0, 0)
+        self.filtered = StatMeasure(0, 0)
+        self.over_quota = StatMeasure(0, 0)
+        self.spike_protection = StatMeasure(0, 0)
+        self.other = StatMeasure(0, 0)
+        self.time = time
+
+    def serialize(self):
+        return {
+            "accepted": {
+                "quantity": self.accepted.quantity,
+                "times_seen": self.accepted.times_seen,
+            },
+            "filtered": {"quantity": self.filtered.quantity, "times_seen": self.filtered.quantity},
+            "dropped": {
+                "overQuota": {
+                    "quantity": self.over_quota.quantity,
+                    "times_seen": self.over_quota.quantity,
+                },
+                "spikeProtection": {
+                    "quantity": self.spike_protection.quantity,
+                    "times_seen": self.spike_protection.quantity,
+                },
+                "other": {"quantity": self.other.quantity, "times_seen": self.other.quantity},
+            },
+            "time": self.time,
+        }
+
+    def update(self, row):
+        if row["outcome"] == Outcome.RATE_LIMITED:
+            if row["reason"] in {"usage_exceeded", "grace_period"}:
+                self.over_quota.quantity = row["quantity"]
+                self.over_quota.times_seen = row["times_seen"]
+            elif row["reason"] in "smart_rate_limit":
+                self.spike_protection.quantity = row["quantity"]
+                self.spike_protection.times_seen = row["quantity"]
+            else:
+                self.other.quantity = row["quantity"]
+                self.other.times_seen = row["quantity"]
+
+        elif row["outcome"] == Outcome.ACCEPTED:
+            self.accepted.quantity = row["quantity"]
+            self.accepted.times_seen = row["times_seen"]
+        elif row["outcome"] == Outcome.FILTERED:
+            self.filtered.quantity = row["quantity"]
+            self.filtered.times_seen = row["times_seen"]
```

Comment:
Suggested refactoring. A few more lines but it trims down the repetition.

```python
    def update(self, row):
        def find_measure():
            if row["outcome"] == Outcome.RATE_LIMITED:
                if row["reason"] in {"usage_exceeded", "grace_period"}:
                    return self.over_quota, True
                elif row["reason"] == "smart_rate_limit":
                    return self.spike_protection, False
                else:
                    return self.other, False
            elif row["outcome"] == Outcome.ACCEPTED:
                return self.accepted, True
            elif row["outcome"] == Outcome.FILTERED:
                return self.filtered, True
            else:
                return None, None  # or raise an error?

        measure_to_update, use_times_seen = find_measure()
        if measure_to_update:
            measure_to_update.quantity = row["quantity"]
            measure_to_update.times_seen = row["times_seen" if use_times_seen else "quantity"]
```

Reply:
👍🏼 will add that in. going to set quantity and times_seen to the row values here and return a singular value in serialization based on the DataCategory, so helper function will just return the one value.